File: C#/Solutions/Problem003/NodeSerializer.cs
namespace DailyCodingProblems.Solutions.Problem003
{
    /// <summary>
    /// Serializes and deserializes all <see cref="Node"/> from a binary tree     
    /// </summary>
    public class NodeSerializer
    {
        private static int _index;

        public static string Serialize(Node node)
        {
            return node == null ? "empty" : $"{node.Value}-{Serialize(node.Left)}-{Serialize(node.Right)}";
        }

        public static Node Deserialize(string serializedNode)
        {
            var nodeList = serializedNode.Split("-");

            if (_index >= nodeList.Length) _index = 0;

            if (nodeList[_index].Contains("empty"))
            {
                _index++;
                return null;
            }

            var nodeToReturn = new Node(nodeList[_index]);
            _index++;
            nodeToReturn.Left = Deserialize(serializedNode);
            nodeToReturn.Right = Deserialize(serializedNode);
            return nodeToReturn;
        }
    }
}

File: Java/Solutions/Problem003/NodeSerializer.java
package Problem003;

/**
 * Serializes and deserializes all Nodes from a binary tree
 */
public class NodeSerializer {
    private static int _index;

    public static String serialize(Node node) {
        return node == null ? "empty" : node.getValue() + "-" + serialize(node.getLeft()) + "-" + serialize(node.getRight());
    }

    public static Node deserialize(String serializedNode) {
        var nodeList = serializedNode.split("-");

        if (_index >= nodeList.length) _index = 0;

        if (nodeList[_index].contains("empty"))
        {
            _index++;
            return null;
        }

        var nodeToReturn = new Node(nodeList[_index], null, null);
        _index++;
        nodeToReturn.setLeft(deserialize(serializedNode));
        nodeToReturn.setRight(deserialize(serializedNode));
        return nodeToReturn;
    }
}
